src/java.base/share/classes/javax/crypto/Cipher.java

```diff
@@ -305,42 +306,67 @@ private static String[] tokenizeTransformation(String transformation)
          * 2) feedback component (e.g., CFB) - optional
          * 3) padding component (e.g., PKCS5Padding) - optional
          */
+        int endIdx = transformation.indexOf('/');
+        if (endIdx == -1) { // done, algo only
+            String algo = transformation.trim();
+            if (algo.isEmpty()) {
+                throw new NoSuchAlgorithmException(
+                        "Invalid transformation: algorithm not specified");
+            }
+            return new String[] { algo };
+        }
 
-        // check if the transformation contains algorithms with "/" in their
-        // name which can cause the parsing logic to go wrong
-        int sha512Idx = transformation.toUpperCase(Locale.ENGLISH)
-                .indexOf(SHA512TRUNCATED);
-        int startIdx = (sha512Idx == -1 ? 0 :
-                sha512Idx + SHA512TRUNCATED.length());
-        int endIdx = transformation.indexOf('/', startIdx);
-
-        boolean algorithmOnly = (endIdx == -1);
-        String algo = (algorithmOnly ? transformation.trim() :
+        // check the presence of special algorithms with "/" in the name, e.g.
+        // truncated SHA512, which can cause the parsing logic to go wrong
+        int sha512Idx = transformation.indexOf(SHA512TRUNCATED);
+        int sha512SlashIdx = (sha512Idx != -1 ? sha512Idx + 3 : -2);
+        // check for the known algorithm slash
+        while (endIdx == sha512SlashIdx) {
+            endIdx = transformation.indexOf('/', sha512SlashIdx + 1);
+            sha512Idx = transformation.indexOf(SHA512TRUNCATED, sha512SlashIdx +
+                    SHA512TRUNCATED.length());
+            sha512SlashIdx = (sha512Idx != -1 ? sha512Idx + 3 : -2);
+        }
```

Comment:
The logic is quite complex. Can you create a dedicated method for it like `indexOfRealSlashIn(String s)`?

Also, although we know there is no `512/2` in valid `mode`, there is no need to treat it specially, especially if we have `indexOfReachSlashIn`.

Reply:
I did refactoring it using a record. If we apply the same method on mode as well, then the flow is simplified even though the mode should not contain truncated algorithms.